geode-core/src/main/java/org/apache/geode/internal/offheap/MemoryAllocatorImpl.java

```diff
@@ -66,7 +67,9 @@ public class MemoryAllocatorImpl implements MemoryAllocator {
 
   private final ScheduledExecutorService updateNonRealTimeStatsExecutor;
 
-  private final ScheduledFuture<?> updateNonRealTimeStatsFuture;
+  private volatile ScheduledFuture<?> updateNonRealTimeStatsFuture;
```

Comment:
The fix I described in the Jira ticket might be too simple. One overall recommendation for newer Java code is to use atomics instead of raw volatile fields. Then you can make it final: `final AtomicReference<Future<?>>`.

Also, I would be tempted to extract `updateNonRealTimeStatsExecutor` and `updateNonRealTimeStatsFuture` together in a new class that `MemoryAllocatorImpl` collaborates with. The smaller and more focused these classes are, the easier it is to test each of them in isolation. A collaborator gives you lots of options for mocking, verifying, etc. I can usually find a class design (ie, composition of multiple classes) that resonates really well and seems to address all of the issues I can find in the code.

Reply:
Looks good.